exporter/SynthesisFusionAddin/src/APS/APS.py

```diff
@@ -0,0 +1,184 @@
+import json
+import logging
+import os
+import pathlib
+import pickle
+import time
+import urllib.parse
+import urllib.request
+from dataclasses import dataclass
+
+from ..general_imports import (
+    APP_NAME,
+    DESCRIPTION,
+    INTERNAL_ID,
+    gm,
+    my_addin_path,
+    root_logger,
+)
+
+CLIENT_ID = "GCxaewcLjsYlK8ud7Ka9AKf9dPwMR3e4GlybyfhAK2zvl3tU"
+auth_path = os.path.abspath(os.path.join(my_addin_path, "..", ".aps_auth"))
+
+APS_AUTH = None
+APS_USER_INFO = None
+
+
+@dataclass
+class APSAuth:
+    access_token: str
+    refresh_token: str
+    expires_in: int
+    expires_at: int
+    token_type: str
+
+
+@dataclass
+class APSUserInfo:
+    name: str
+    given_name: str
+    family_name: str
+    preferred_username: str
+    email: str
+    email_verified: bool
+    profile: str
+    locale: str
+    country_code: str
+    about_me: str
+    language: str
+    company: str
+    picture: str
+
+
+def getAPSAuth() -> APSAuth | None:
+    return APS_AUTH
+
+
+def _res_json(res):
+    return json.loads(res.read().decode(res.info().get_param("charset") or "utf-8"))
+
+
+def getCodeChallenge() -> str | None:
+    endpoint = 'http://localhost:3003/api/aps/challenge/'
+    res = urllib.request.urlopen(endpoint)
+    data = _res_json(res)
+    return data["challenge"]
+
+
+def getAuth() -> APSAuth:
+    global APS_AUTH
+    if APS_AUTH is not None:
+        return APS_AUTH
+    try:
+        with open(auth_path, "rb") as f:
+            p = pickle.load(f)
+            APS_AUTH = APSAuth(
+                access_token=p["access_token"],
+                refresh_token=p["refresh_token"],
+                expires_in=p["expires_in"],
+                expires_at=int(p["expires_in"] * 1000),
+                token_type=p["token_type"],
+            )
+    except:
+        raise Exception("Need to sign in!")
+    curr_time = int(time.time() * 1000)
+    if curr_time >= APS_AUTH.expires_at:
+        refreshAuthToken()
+    if APS_USER_INFO is None:
+        loadUserInfo()
+    return APS_AUTH
+
+
+def convertAuthToken(code: str):
+    global APS_AUTH
+    authUrl = f'http://localhost:3003/api/aps/code/?code={code}&redirect_uri={urllib.parse.quote_plus("http://localhost:3003/api/aps/exporter/")}'
```

Comment:
These will have to be changed to the actually URL eventually, right? A constant might be helpful for this.